### ISSUE TITLE
ArrayIndexOutOfBounds exception in BinaryFrameHeader

### ISSUE THREAD

**Issue by jthrun** [smartdevicelink/sdl_android#332]
This happened once, so far, and I can't repro it but it caused the proxy to cycle.  I don't know what the expected behavior in this case is, cycling the proxy or handling the case where there might be invalid data.  Unfortunately, I did not save the stack trace, but in any case, the exception happened here:
https://github.com/smartdevicelink/sdl_android/blob/develop/sdl_android_lib/src/com/smartdevicelink/protocol/BinaryFrameHeader.java#L33

``` java
        if (_jsonSize > 0) {
            byte[] _jsonData = new byte[_jsonSize];
            System.arraycopy(binHeader, 12, _jsonData, 0, _jsonSize);    // Line 33
            msg.setJsonData(_jsonData);
        }
```

**Comment by joeygrover** [comment]
The only guess I would have is it being due to a v1 packet trying to be processed in the router service which would only have 8 bytes instead of 12 for the packet header. Every other place else it seems to be checking to see if it's a v1 packet before it uses that method. I have a local branch that should check, but I'm not sure it's the issue.

https://github.com/smartdevicelink/sdl_android/blob/master/sdl_android_lib/src/com/smartdevicelink/transport/SdlRouterService.java#L1371

Let me know if you get a stack trace.

**Comment by jthrun** [comment]
Would 4.1 release of SDL core send a v1 packet?  What might trigger this to happen?

**Comment by joeygrover** [comment]
The initial StartServiceAck from Core for the RPC Service is technically a v1 packet. This is to maintain backwards compatibility for both mobile and core. This is because v1 packets didn't have a message id section in their header (4 bytes right before the data piece starts). So this is where the out of range error could come from.

**Comment by jthrun** [comment]
For some reason, today I am getting this error now almost every time I change languages on the HU.  Here is the stack trace:
```
10-12 17:27:21.522  5684  6436 E SdlProxy: java.lang.ArrayIndexOutOfBoundsException: src.length=10670 srcPos=12 dst.length=12146 dstPos=0 length=12146
10-12 17:27:21.522  5684  6436 E SdlProxy:      at java.lang.System.arraycopy(Native Method)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at com.smartdevicelink.protocol.BinaryFrameHeader.parseBinaryHeader(BinaryFrameHeader.java:33)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at com.smartdevicelink.protocol.WiProProtocol$MessageFrameAssembler.notifyIfFinished(WiProProtocol.java:280)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at com.smartdevicelink.protocol.WiProProtocol$MessageFrameAssembler.handleRemainingFrame(WiProProtocol.java:267)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at com.smartdevicelink.protocol.WiProProtocol$MessageFrameAssembler.handleMultiFrameMessageFrame(WiProProtocol.java:310)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at com.smartdevicelink.protocol.WiProProtocol$MessageFrameAssembler.handleFrame(WiProProtocol.java:346)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at com.smartdevicelink.protocol.WiProProtocol.handlePacketReceived(WiProProtocol.java:227)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at com.smartdevicelink.SdlConnection.SdlConnection.onTransportPacketReceived(SdlConnection.java:207)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at com.smartdevicelink.transport.SdlTransport.handleReceivedPacket(SdlTransport.java:42)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at com.smartdevicelink.transport.MultiplexTransport$TransportBrokerThread$1.onPacketReceived(MultiplexTransport.java:295)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at com.smartdevicelink.transport.TransportBroker$ClientHandler.handleMessage(TransportBroker.java:213)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at android.os.Handler.dispatchMessage(Handler.java:102)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at android.os.Looper.loop(Looper.java:135)
10-12 17:27:21.522  5684  6436 E SdlProxy:      at com.smartdevicelink.transport.MultiplexTransport$TransportBrokerThread.run(MultiplexTransport.java:230)
```

It happens to all the apps that are registered but the sizes are not the same in each app.  Here's another example:

```
10-12 17:27:21.329  5553  6435 E SdlProxy: java.lang.ArrayIndexOutOfBoundsException: src.length=10531 srcPos=12 dst.length=13495 dstPos=0 length=13495
```

**Comment by jthrun** [comment]
Note, the HU has the 4.0 version of SDLCore.

**Comment by joeygrover** [comment]
Ok based on that log it's not what I thought. I will take a look at this

**Comment by askirk** [comment]
@jthrun Can you print out the functionID of the message that it fails on? Also, do you know what exact version of SDL Core you're using (4.0.X)? I'm trying to put together a timeline to reproduce this.

**Comment by jthrun** [comment]
Is that information available at the point where the exception occurred?

Not sure about the exact version but it was 4.0.10 or 4.0.12.

**Comment by askirk** [comment]
It should be. Try running in Debug mode, or you can log it after this line and find out the last functionID received before the error.

https://github.com/smartdevicelink/sdl_android/blob/develop/sdl_android_lib/src/com/smartdevicelink/protocol/BinaryFrameHeader.java#L22

**Comment by jthrun** [comment]
OK, I'll give it a shot.

**Comment by askirk** [comment]
Any update?

**Comment by jthrun** [comment]
No, we've been working on releases.  I'll try to repro today or tomorrow.

**Comment by jthrun** [comment]
I am unable to reproduce this on two of the three HUs we have.  On the third, I can't run SDL apps so I can't perform a language change anyway.  It can't be repro'd on our dev boards because they don't have language change capability.  So, for today anyway, I'm stuck.

**Comment by jthrun** [comment]
@askirk I was capturing logs yesterday for a different reason and ran into several of these errors.  I am thinking the cause may be that the HU dropped some BT packets but I am not sure.  In any case here are several instances of the error from two different devices:

Device 1 - Samsung Galaxy S5 - OS v6.0.1
```
04-27 17:24:38.185  8649 11061 V JT_DBG_BFH: '12950' BinaryFrameHeader [type=1, functionID=1, corrID=65529, JSON size=12223] size: 10747
04-27 17:24:38.185  8649 11061 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=10747 srcPos=12 dst.length=12223 dstPos=0 length=12223
```
```
04-27 17:40:02.570  8887 11569 V JT_DBG_BFH: '12006' BinaryFrameHeader [type=2, functionID=32776, corrID=-1, JSON size=13189] size: 11713
04-27 17:40:02.570  8887 11569 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=11713 srcPos=12 dst.length=13189 dstPos=0 length=13189
```
```
04-27 17:40:05.490  8887 12148 V JT_DBG_BFH: '12006' BinaryFrameHeader [type=2, functionID=32776, corrID=-1, JSON size=13189] size: 11713
04-27 17:40:05.490  8887 12148 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=11713 srcPos=12 dst.length=13189 dstPos=0 length=13189
```

Device 2 - Nexus 5 - OS v5.0.1
```
04-27 17:23:22.883  5460  6636 V JT_DBG_BFH: '12004' BinaryFrameHeader [type=2, functionID=32776, corrID=-1, JSON size=13189] size: 11713
04-27 17:23:22.883  5460  6636 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=11713 srcPos=12 dst.length=13189 dstPos=0 length=13189
```
```
04-27 17:23:23.490  5398  6641 V JT_DBG_BFH: '12006' BinaryFrameHeader [type=1, functionID=1, corrID=65529, JSON size=12223] size: 9259
04-27 17:23:23.490  5398  6641 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=9259 srcPos=12 dst.length=12223 dstPos=0 length=12223
```
```
04-27 17:26:52.884  5460  7705 V JT_DBG_BFH: '12004' BinaryFrameHeader [type=1, functionID=1, corrID=65529, JSON size=12223] size: 10747
04-27 17:26:52.884  5460  7705 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=10747 srcPos=12 dst.length=12223 dstPos=0 length=12223
```
```
04-27 17:26:53.941  5271  7714 V JT_DBG_BFH: '12002' BinaryFrameHeader [type=1, functionID=1, corrID=65529, JSON size=12223] size: 10747
04-27 17:26:53.941  5271  7714 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=10747 srcPos=12 dst.length=12223 dstPos=0 length=12223
```
```
04-27 17:26:55.446  5511  7701 V JT_DBG_BFH: '12005' BinaryFrameHeader [type=2, functionID=32776, corrID=-1, JSON size=13462] size: 11986
04-27 17:26:55.446  5511  7701 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=11986 srcPos=12 dst.length=13462 dstPos=0 length=13462
```
```
04-27 17:26:59.993  5511  7952 V JT_DBG_BFH: '12005' BinaryFrameHeader [type=1, functionID=1, corrID=65529, JSON size=12223] size: 10747
04-27 17:26:59.993  5511  7952 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=10747 srcPos=12 dst.length=12223 dstPos=0 length=12223
```
```
04-27 17:36:07.804  5460  8578 V JT_DBG_BFH: '12004' BinaryFrameHeader [type=2, functionID=32776, corrID=-1, JSON size=13189] size: 11713
04-27 17:36:07.804  5460  8578 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=11713 srcPos=12 dst.length=13189 dstPos=0 length=13189
```
```
04-27 17:36:09.410  5511  8574 V JT_DBG_BFH: '12005' BinaryFrameHeader [type=1, functionID=1, corrID=65529, JSON size=12223] size: 10747
04-27 17:36:09.410  5511  8574 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=10747 srcPos=12 dst.length=12223 dstPos=0 length=12223
```
```
04-27 17:36:10.675  5336  8584 V JT_DBG_BFH: '12008' BinaryFrameHeader [type=1, functionID=1, corrID=65529, JSON size=12223] size: 10747
04-27 17:36:10.675  5336  8584 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=10747 srcPos=12 dst.length=12223 dstPos=0 length=12223
```
```
04-27 17:36:12.005  5215  8592 V JT_DBG_BFH: '12950' BinaryFrameHeader [type=1, functionID=1, corrID=65529, JSON size=12223] size: 10747
04-27 17:36:12.005  5215  8592 V JT_DBG_BFH: java.lang.ArrayIndexOutOfBoundsException: src.length=10747 srcPos=12 dst.length=12223 dstPos=0 length=12223
```

**Comment by jthrun** [comment]
Regardless of the cause, i.e. dropped packets or whatever, wouldn't it be better for `BinaryFrameHeader.parseBinaryHeader` to return null rather than letting it crash the proxy?

**Comment by askirk** [comment]
Yes, I'll add a catch clause for this in an pull request

**Comment by joeygrover** [comment]
Code fix has been merged into develop through #478. Will be contained in next release.
